sympy/core/expr.py

```diff
@@ -146,9 +146,25 @@ def __rmul__(self, other):
 
     @_sympifyit('other', NotImplemented)
     @call_highest_priority('__rpow__')
-    def __pow__(self, other):
+    def _pow(self, other):
         return Pow(self, other)
 
+    def __pow__(self, other, mod=None):
+        if mod is None:
+            return self._pow(other)
+        try:
+            _self, other, mod = as_int(self), as_int(other), as_int(mod)
+            if other >= 0:
+                return pow(_self, other, mod)
+            else:
+                from sympy.core.numbers import mod_inverse
+                return mod_inverse(pow(_self, -other, mod), mod)
+        except ValueError:
+            power = self._pow(other)
+            if power is NotImplemented:
```

Comment:
Can this really be true?

Reply:
@jksuom `_pow` returns `NotImplemented` when _sympifying `other` fails, e.g., `pow(S(2), "2", 3)`. If we don't catch it, as `%`(mod) operation is not defined for `NotImplementedType`, it raises `TypeError: unsupported operand type(s) for %: 'NotImplementedType' and 'type(mod)'`, which I think is confusing and less informative. If we catch it and return `NotImplemented`, python raises `TypeError: unsupported operand type(s) for pow(): 'Integer', 'str', 'int'`.

Reply:
> it raises TypeError: unsupported operand type(s)

Then should the code be
```python
try:
    power = self._pow(other)
except TypeError:
    raise NotImplementedError  <--note NotImplementedError instead of NotImplemented?
```

And a test like `raises(NotImplementedError, lambda: pow(...failing args...))` should be added.

Reply:
@smichr `_pow` itself does not raises TypeError, rather it returns `NotImplemented` when _sympify fails. Python automatically constructs and raises `TypeError` if `NotImplemented` is returned by overridden default operations, e.g., `__pow__`.

Reply:
> _pow returns NotImplemented when _sympifying other fails, 

I didn't know that. I thought that `SympifyError` or `NotImplementedError` was raised if `other` is not sympifiable in `Pow(self, other)`.

Reply:
Can we get a test for this added to the suite so this line is covered?

Reply:
@smichr Added 😄

Reply:
Changed to catch TypeError, for when `mod` is not _sympifiable.